src/crimson/os/seastore/journal.cc

```diff
@@ -720,4 +734,110 @@ Journal::scan_segment_ret Journal::scan_segment(
     });
 }
 
+Journal::scan_valid_records_ret Journal::scan_valid_records(
+    scan_valid_records_cursor &cursor,
+    segment_nonce_t nonce,
+    size_t budget,
+    found_record_handler_t &handler)
+{
+  if (cursor.offset.offset == 0) {
+    cursor.offset.offset = block_size;
+  }
+  auto retref = std::make_unique<size_t>(0);
+  auto budget_used = *retref;
+  return crimson::do_until(
+    [=, &cursor, &budget_used, &handler]() mutable
+    -> scan_valid_records_ertr::future<bool> {
+      return [=, &handler, &cursor, &budget_used] {
+	if (!cursor.last_valid_header_found) {
+	  return read_validate_record_metadata(cursor.offset, nonce
+	  ).safe_then([=, &cursor, &budget_used, &handler](auto md) {
+	    logger().debug(
+	      "Journal::scan_valid_records: read complete {}",
+	      cursor.offset);
+	    if (!md) {
+	      logger().debug(
+		"Journal::scan_valid_records: found invalid header at {}",
+		cursor.offset);
+	      cursor.last_valid_header_found = true;
+	      return scan_valid_records_ertr::now();
+	    } else {
+	      logger().debug(
+		"Journal::scan_valid_records: valid record read at {}",
+		cursor.offset);
+	      cursor.last_committed = paddr_t{
+		cursor.offset.segment,
+		md->first.committed_to};
```

Comment:
I don't understand why need to persist `committed_to` in each `record_header_t`. According to implementation in https://github.com/ceph/ceph/pull/38290/commits/83414be599cd40af8848958cd502e720d4d04e84, it should be always pointing to the start offset of this record, correct?

Records are written in order, if we can make sure the next record's metadata is valid through nonce and crc checks, the current record should be committed already, right?

Reply:
No, the update to committed_to happens after the record commits.  Any record gets encoded with the address of the latest record to commit, which will be at least one record back and in the event that we've pipelined many of them possible more.  The purpose of committed_to is an optimization to avoid calculating data_crc except on the last few records of a segment.  If we only used this for replay, I wouldn't have bothered.  However, gc uses the same machinery, so it's in the fast path.

Reply:
Regardless of whether we've pipelined multiple records to submit or not. I'm wondering if the following pseudo code is identical to the current `scan_valid_records()` which can bypass most of data_crc checks and can work without `committed_to`:
```
INPUT: cursor_offset, nonce, budget, handler

current_md <- nil;
next_md <- nil;
budget_used <- 0;
if (cursor_offset == 0) {
    cursor_offset <- block_size
}
while (true) {
    current_md <- next_md
    next_md <- read_validate_record_metadata(cursor_offset, nonce)
    if (!next_md) {
        // the metadata of the next record is invalid, stop scanning
        break;
    } else if (current_md) {
        // the metadata of the next record is valid,
        // meaning the current record has completed submission,
        // thus no need to check data crc.
        handler(cursor_offset, current_md.header, current_md.mdbuffer)
        if (budget_used >= budget) {
            return budget_used
        }
    }
    record_length <- next_md.header.mdlength + next_md.header.dlength
    budget_used <- budget_used + record_length
    cursor_offset <- cursor_offset + record_length
}
if (current_md) {
    // the metadata of the next record is invalid,
    // meaning the current record is the final record and needs to check data crc.
    is_data_valid <- read_validate_data(cursor_offset, current_md.header)
    if (is_data_valid) {
        handler(cursor_offset, current_md.header, current_md.mdbuffer)
        record_length <- current_md.header.mdlength + current_md.header.dlength
        budget_used <- budget_used + record_length
    } else {
        // the current record is not valid, ignore
    }
} else {
    // the segment is empty.
}
return budget_used
```

Reply:
Your pseudocode works under the assumption that

   valid_metadata_crc_record(N+1) => valid_record(N)

This won't be true in general.  If we have N...N+5 in flight at the same time, the disk may persist the metadata for all of them without persisting the data for any of them.  The disk itself has no such ordering guarantee, and we don't necessarily wait for a flush on N before submitting N+1.

Reply:
Thanks, it makes sense.